### ISSUE TITLE
Question regarding the persist library

### ISSUE THREAD

**Issue by minad** [ollef/sixty#3]
Hi @ollef,

I am following this project with interest and saw that you are using my [persist](https://github.com/minad/persist/) library to serialize the AST. I am also working on a language project, where I am serializing AST datastructures. So I thought I may ask for some feedback regarding serialization/improvements of the persist library.

There is one issue I had - I would like to deduplicate certain parts of the AST, e.g. names. Therefore I experimented with a version of [persist with state](https://github.com/minad/persist-state/), which augments the Get and Put monad with additional state such that a HashMap of already serialized names can be passed around. This helps at least a bit to make the serialized data a bit more compact. I didn't test how much this affects performance and I am not quite sure if this is a good approach. Do you have similar problems or do you serialize keep de-bruijn indexed datastructures and keep names separately in a symbol table? Maybe there are other better approaches, e.g. manually writing Get/Put instances in order to transform everything to the desired format etc or maybe use some other intermediate datastructures with generated instances, but this sounds slow.

Daniel

**Comment by ollef** [comment]
Hey Daniel,

To be honest I haven't yet started using the persist library to actually persist anything to disk, so I don't have any great feedback, but it seems very plausible that deduplicating names would save a lot of space in Sixty, since top-level definitions are referred to by their full, qualified name.

I've experimented a bit with name interning, and a loose thought is that perhaps there's some integration that can be made between persist and such a library.

**Comment by minad** [comment]
> but it seems very plausible that deduplicating names would save a lot of space in Sixty, since top-level definitions are referred to by their full, qualified name.

Ok, in this case persist-state might help!

> I've experimented a bit with name interning, and a loose thought is that perhaps there's some integration that can be made between persist and such a library.

I also experimented with interning but in a quite naive way (global HashMap, modified in unsafePerformIO, everything interned leaks). I am using interning more to achieve fast equality checking, similar to what GHC does with FastString. 

I also thought about coupling this somehow with the persistence mechanism, but didn't come up with a good solution yet. Maybe traverse the AST once, serialize all used names as some kind of table. Then serialize the AST and write the interned ids. But for reading I still need some kind of state since the interned ids are generated by the interning mechanism. Maybe one could do something more sophisticated using a table associated with the AST, which contains only the interned names corresponding to the AST and not this global interning that I am playing with right now. But this requires some effort since you also have to pass around the table with the AST everywhere.

So in the end I modified persist such that it allows state. I am wondering why no other serialization library has thought about this issue. Maybe people use monad transformers on top of binary/cereal/store etc. But this might have quite some cost.
